diff --git a/sbin/growfs/growfs.c b/sbin/growfs/growfs.c
index daa3931df30..f02db936b24 100644
--- a/sbin/growfs/growfs.c
+++ b/sbin/growfs/growfs.c
@@ -733,7 +733,7 @@ updjcg(int cylno, time_t modtime, int fsi, int fso, unsigned int Nflag)
 	}
 
 	/*
-	 * Handle the last new block if there are stll some new fragments left.
+	 * Handle the last new block if there are still some new fragments left.
 	 * Here we don't have to bother about the cluster summary or the even
 	 * the rotational layout table.
 	 */
