diff --git a/sys/compat/cheriabi/cheriabi_syscalls.c b/sys/compat/cheriabi/cheriabi_syscalls.c
index d45dc78..a36a2ed8 100644
--- a/sys/compat/cheriabi/cheriabi_syscalls.c
+++ b/sys/compat/cheriabi/cheriabi_syscalls.c
@@ -349,7 +349,7 @@ const char *cheriabi_syscallnames[] = {
 	"obs_sendfile",			/* 336 = obsolete sendfile */
 	"kldsym",			/* 337 = kldsym */
 	"cheriabi_jail",			/* 338 = cheriabi_jail */
-	"#339",			/* 339 = pioctl */
+	"#339",			/* 339 = nnpfs_syscall */
 	"sigprocmask",			/* 340 = sigprocmask */
 	"sigsuspend",			/* 341 = sigsuspend */
 	"obs_sigaction",			/* 342 = obsolete sigaction */
@@ -387,7 +387,7 @@ const char *cheriabi_syscallnames[] = {
 	"__setugid",			/* 374 = __setugid */
 	"#375",			/* 375 = nfsclnt */
 	"eaccess",			/* 376 = eaccess */
-	"#377",			/* 377 = afs_syscall */
+	"#377",			/* 377 = afs3_syscall */
 	"cheriabi_nmount",			/* 378 = cheriabi_nmount */
 	"#379",			/* 379 = kse_exit */
 	"#380",			/* 380 = kse_wakeup */
diff --git a/sys/compat/cheriabi/cheriabi_sysent.c b/sys/compat/cheriabi/cheriabi_sysent.c
index 7a29a36..4bbc749 100644
--- a/sys/compat/cheriabi/cheriabi_sysent.c
+++ b/sys/compat/cheriabi/cheriabi_sysent.c
@@ -360,7 +360,7 @@ struct sysent cheriabi_sysent[] = {
 	{ 0, (sy_call_t *)nosys, AUE_NULL, NULL, 0, 0, 0, SY_THR_ABSENT },			/* 336 = obsolete sendfile */
 	{ AS(kldsym_args), (sy_call_t *)sys_kldsym, AUE_NULL, NULL, 0, 0, 0, SY_THR_STATIC },	/* 337 = kldsym */
 	{ AS(cheriabi_jail_args), (sy_call_t *)cheriabi_jail, AUE_JAIL, NULL, 0, 0, 0, SY_THR_STATIC },	/* 338 = cheriabi_jail */
-	{ 0, (sy_call_t *)nosys, AUE_NULL, NULL, 0, 0, 0, SY_THR_ABSENT },			/* 339 = pioctl */
+	{ 0, (sy_call_t *)nosys, AUE_NULL, NULL, 0, 0, 0, SY_THR_ABSENT },			/* 339 = nnpfs_syscall */
 	{ AS(sigprocmask_args), (sy_call_t *)sys_sigprocmask, AUE_SIGPROCMASK, NULL, 0, 0, SYF_CAPENABLED, SY_THR_STATIC },	/* 340 = sigprocmask */
 	{ AS(sigsuspend_args), (sy_call_t *)sys_sigsuspend, AUE_SIGSUSPEND, NULL, 0, 0, SYF_CAPENABLED, SY_THR_STATIC },	/* 341 = sigsuspend */
 	{ 0, (sy_call_t *)nosys, AUE_NULL, NULL, 0, 0, 0, SY_THR_ABSENT },			/* 342 = obsolete sigaction */
@@ -398,7 +398,7 @@ struct sysent cheriabi_sysent[] = {
 	{ AS(__setugid_args), (sy_call_t *)sys___setugid, AUE_NULL, NULL, 0, 0, 0, SY_THR_STATIC },	/* 374 = __setugid */
 	{ 0, (sy_call_t *)nosys, AUE_NULL, NULL, 0, 0, 0, SY_THR_ABSENT },			/* 375 = nfsclnt */
 	{ AS(eaccess_args), (sy_call_t *)sys_eaccess, AUE_EACCESS, NULL, 0, 0, 0, SY_THR_STATIC },	/* 376 = eaccess */
-	{ 0, (sy_call_t *)nosys, AUE_NULL, NULL, 0, 0, 0, SY_THR_ABSENT },			/* 377 = afs_syscall */
+	{ 0, (sy_call_t *)nosys, AUE_NULL, NULL, 0, 0, 0, SY_THR_ABSENT },			/* 377 = afs3_syscall */
 	{ AS(cheriabi_nmount_args), (sy_call_t *)cheriabi_nmount, AUE_NMOUNT, NULL, 0, 0, 0, SY_THR_STATIC },	/* 378 = cheriabi_nmount */
 	{ 0, (sy_call_t *)nosys, AUE_NULL, NULL, 0, 0, 0, SY_THR_ABSENT },			/* 379 = kse_exit */
 	{ 0, (sy_call_t *)nosys, AUE_NULL, NULL, 0, 0, 0, SY_THR_ABSENT },			/* 380 = kse_wakeup */
