$FreeBSD$

Index: contrib/gcc/config/rs6000/freebsd.h
===================================================================
RCS file: /usr/home/ncvs/src/contrib/gcc/config/rs6000/freebsd.h,v
retrieving revision 1.1.1.2
diff -u -r1.1.1.2 freebsd.h
--- contrib/gcc/config/rs6000/freebsd.h	11 Jul 2003 03:40:53 -0000	1.1.1.2
+++ contrib/gcc/config/rs6000/freebsd.h	15 Jul 2003 22:17:12 -0000
@@ -43,6 +43,23 @@
 #undef  LINK_SHLIB_SPEC
 #define LINK_SHLIB_SPEC "%{shared:-shared} %{!shared: %{static:-static}}"
 
+/* We don't want _BIG_ENDIAN or _LITTLE_ENDIAN to be defined. */
+#undef RS6000_CPU_CPP_ENDIAN_BUILTINS
+#define RS6000_CPU_CPP_ENDIAN_BUILTINS()	\
+  do						\
+    {						\
+      if (BYTES_BIG_ENDIAN)			\
+	{					\
+	  builtin_define ("__BIG_ENDIAN__");	\
+	  builtin_assert ("machine=bigendian");	\
+	}					\
+      else					\
+	{					\
+	  builtin_define ("__LITTLE_ENDIAN__");	\
+	  builtin_assert ("machine=littleendian"); \
+	}					\
+    }						\
+  while (0)
 
 /************************[  Target stuff  ]***********************************/
 
Index: contrib/gcc/config/rs6000/sysv4.h
===================================================================
RCS file: /usr/home/ncvs/src/contrib/gcc/config/rs6000/sysv4.h,v
retrieving revision 1.1.1.6
diff -u -r1.1.1.6 sysv4.h
--- contrib/gcc/config/rs6000/sysv4.h	11 Jul 2003 03:40:53 -0000	1.1.1.6
+++ contrib/gcc/config/rs6000/sysv4.h	15 Jul 2003 20:35:39 -0000
@@ -21,8 +21,6 @@
 Boston, MA 02111-1307, USA.  */
 
 
-/* Header files should be C++ aware in general.  */
-#define NO_IMPLICIT_EXTERN_C
 
 /* Yes!  We are ELF.  */
 #define	TARGET_OBJECT_FORMAT OBJECT_ELF
Index: gnu/usr.bin/cc/cc_int/Makefile
===================================================================
RCS file: /usr/home/ncvs/src/gnu/usr.bin/cc/cc_int/Makefile,v
retrieving revision 1.34
diff -u -r1.34 Makefile
--- gnu/usr.bin/cc/cc_int/Makefile	11 Jul 2003 05:37:23 -0000	1.34
+++ gnu/usr.bin/cc/cc_int/Makefile	15 Jul 2003 21:26:53 -0000
@@ -27,6 +27,10 @@
 	attribs.c cselib.c debug.c rtl-error.c tree-dump.c tree-inline.c
 
 SRCS+=	${GCC_CPU}.c
+.if exists(${GCC_CPU}-c.c)
+SRCS+=	${GCC_CPU}-c.c
+.endif
+
 
 SRCS+=	bb-reorder.c conflict.c ggc-common.c \
 	ggc-page.c ifcvt.c lists.c predict.c regrename.c resource.c sibcall.c \
