$FreeBSD$

Index: contrib/gcc/config/rs6000/freebsd.h
===================================================================
RCS file: /home/ncvs/src/contrib/gcc/config/rs6000/freebsd.h,v
retrieving revision 1.1.1.2
diff -u -r1.1.1.2 freebsd.h
--- contrib/gcc/config/rs6000/freebsd.h	11 Jul 2003 03:40:53 -0000	1.1.1.2
+++ contrib/gcc/config/rs6000/freebsd.h	9 Sep 2003 06:51:49 -0000
@@ -43,6 +43,23 @@
 #undef  LINK_SHLIB_SPEC
 #define LINK_SHLIB_SPEC "%{shared:-shared} %{!shared: %{static:-static}}"
 
+/* We don't want _BIG_ENDIAN or _LITTLE_ENDIAN to be defined. */
+#undef RS6000_CPU_CPP_ENDIAN_BUILTINS
+#define RS6000_CPU_CPP_ENDIAN_BUILTINS()	\
+  do						\
+    {						\
+      if (BYTES_BIG_ENDIAN)			\
+	{					\
+	  builtin_define ("__BIG_ENDIAN__");	\
+	  builtin_assert ("machine=bigendian");	\
+	}					\
+      else					\
+	{					\
+	  builtin_define ("__LITTLE_ENDIAN__");	\
+	  builtin_assert ("machine=littleendian"); \
+	}					\
+    }						\
+  while (0)
 
 /************************[  Target stuff  ]***********************************/
 
@@ -58,6 +75,9 @@
 
 #undef  WCHAR_TYPE_SIZE
 #define WCHAR_TYPE_SIZE 32
+
+#undef  DEFAULT_SIGNED_CHAR
+#define DEFAULT_SIGNED_CHAR 1
 
 #undef  TARGET_VERSION
 #define TARGET_VERSION fprintf (stderr, " (FreeBSD/PowerPC ELF)");
Index: contrib/gcc/config/rs6000/sysv4.h
===================================================================
RCS file: /home/ncvs/src/contrib/gcc/config/rs6000/sysv4.h,v
retrieving revision 1.1.1.6
diff -u -r1.1.1.6 sysv4.h
--- contrib/gcc/config/rs6000/sysv4.h	11 Jul 2003 03:40:53 -0000	1.1.1.6
+++ contrib/gcc/config/rs6000/sysv4.h	9 Sep 2003 06:51:49 -0000
@@ -21,8 +21,6 @@
 Boston, MA 02111-1307, USA.  */
 
 
-/* Header files should be C++ aware in general.  */
-#define NO_IMPLICIT_EXTERN_C
 
 /* Yes!  We are ELF.  */
 #define	TARGET_OBJECT_FORMAT OBJECT_ELF
