--- linux.h.orig	Tue Dec  3 16:47:28 1996
+++ linux.h	Mon Nov 17 00:05:29 1997
@@ -491,6 +491,9 @@
 #define LINUX_SNDCTL_DSP_GETOSPACE	0x500C
 #define LINUX_SNDCTL_DSP_GETISPACE	0x500D
 #define LINUX_SNDCTL_DSP_NONBLOCK	0x500E
+#define LINUX_SNDCTL_DSP_GETCAPS	0x500F
+#define LINUX_SNDCTL_DSP_GETIPTR        0x5011
+#define LINUX_SNDCTL_DSP_GETOPTR        0x5012
 #define LINUX_SOUND_MIXER_WRITE_VOLUME	0x4d00
 #define LINUX_SOUND_MIXER_WRITE_BASS	0x4d01
 #define LINUX_SOUND_MIXER_WRITE_TREBLE	0x4d02
--- linux_ioctl.c.orig	Sat Nov  9 22:10:15 1996
+++ linux_ioctl.c	Mon Nov 17 10:20:14 1997
@@ -691,6 +691,26 @@
 	args->cmd = SNDCTL_DSP_NONBLOCK;
 	return ioctl(p, (struct ioctl_args *)args, retval);
 
+    case LINUX_SNDCTL_DSP_GETCAPS:
+	args->cmd = SNDCTL_DSP_GETCAPS;
+	return ioctl(p, (struct ioctl_args *)args, retval);
+
+    case LINUX_SNDCTL_DSP_GETIPTR:
+	args->cmd = SNDCTL_DSP_GETIPTR;
+	return ioctl(p, (struct ioctl_args *)args, retval);
+
+    case LINUX_SNDCTL_DSP_GETOPTR:
+	args->cmd = SNDCTL_DSP_GETOPTR;
+	{ int a= ioctl(p, (struct ioctl_args *)args, retval);
+	struct count_info *p = (struct count_info *)(args->arg);
+#if 0
+       p->bytes += 128 ;
+	uprintf("GETOPTR bytes %d blk %d ptr %d\n",
+		p->bytes, p->blocks, p->ptr);
+#endif
+	return a;
+	}
+
     case LINUX_SOUND_MIXER_WRITE_VOLUME:
 	args->cmd = SOUND_MIXER_WRITE_VOLUME;
 	return ioctl(p, (struct ioctl_args *)args, retval);
