*** Makefile.orig	Sat Aug  1 20:35:53 1998
--- Makefile	Sat Aug  1 20:36:50 1998
***************
*** 9,22 ****
  #Uncomment to activate login_auth
  #Warning: requires src/libexec/login_* auth modules
  #LC_AUTH=-DLOGIN_CAP_AUTH
! CFLAGS+=-Wall -DSKEY -DLOGIN_ACCESS -DLOGALL -DLOGIN_CAP $(LC_AUTH)
  
  .if defined(KLOGIN_PARANOID)
  CFLAGS+=-DKLOGIN_PARANOID
  .endif
  
! DPADD=	${LIBUTIL} ${LIBSKEY} ${LIBMD} ${LIBCRYPT}
! LDADD=	-lutil -lskey -lmd -lcrypt
  
  .if exists(${DESTDIR}/usr/lib/libkrb.a) && defined(MAKE_KERBEROS4) && !defined(LC_AUTH)
  CFLAGS+=-DKERBEROS
--- 9,22 ----
  #Uncomment to activate login_auth
  #Warning: requires src/libexec/login_* auth modules
  #LC_AUTH=-DLOGIN_CAP_AUTH
! CFLAGS+=-Wall $(LC_AUTH)
  
  .if defined(KLOGIN_PARANOID)
  CFLAGS+=-DKLOGIN_PARANOID
  .endif
  
! DPADD=	${LIBUTIL} ${LIBMD} ${LIBCRYPT}
! LDADD=	-lutil -lmd -lcrypt
  
  .if exists(${DESTDIR}/usr/lib/libkrb.a) && defined(MAKE_KERBEROS4) && !defined(LC_AUTH)
  CFLAGS+=-DKERBEROS
