*** Makefile.orig	Sat Aug  1 20:39:04 1998
--- Makefile	Sat Aug  1 20:40:38 1998
***************
*** 2,24 ****
  # $Id: passwd.diff,v 1.1 1998/08/19 06:31:06 abial Exp $
  
  PROG=	passwd
! SRCS=	local_passwd.c passwd.c pw_copy.c pw_util.c pw_yp.c \
! 	yp_passwd.c ypxfr_misc.c ${GENSRCS}
  GENSRCS=yp.h yp_clnt.c yppasswd.h yppasswd_clnt.c \
  	yppasswd_private.h yppasswd_private_clnt.c yppasswd_private_xdr.c
  CFLAGS+=-Wall
  
! DPADD=	${LIBCRYPT} ${LIBRPCSVC} ${LIBUTIL}
! LDADD=	-lcrypt -lrpcsvc -lutil
  .PATH:  ${.CURDIR}/../../usr.bin/chpass ${.CURDIR}/../../usr.sbin/vipw \
! 	${.CURDIR}/../rlogin ${.CURDIR}/../../libexec/ypxfr \
! 	${.CURDIR}/../../usr.sbin/rpc.yppasswdd
  
! CFLAGS+= -DLOGIN_CAP -DCRYPT -DYP -I. -I${.CURDIR} \
  	-I${.CURDIR}/../../usr.sbin/vipw \
  	-I${.CURDIR}/../../usr.bin/chpass \
- 	-I${.CURDIR}/../../libexec/ypxfr \
- 	-I${.CURDIR}/../../usr.sbin/rpc.yppasswdd \
  	-Dyp_error=warnx -DLOGGING
  
  CLEANFILES= ${GENSRCS}
--- 2,21 ----
  # $Id: passwd.diff,v 1.1 1998/08/19 06:31:06 abial Exp $
  
  PROG=	passwd
! SRCS=	local_passwd.c passwd.c pw_copy.c pw_util.c 
! 
  GENSRCS=yp.h yp_clnt.c yppasswd.h yppasswd_clnt.c \
  	yppasswd_private.h yppasswd_private_clnt.c yppasswd_private_xdr.c
  CFLAGS+=-Wall
  
! DPADD=	${LIBCRYPT} ${LIBUTIL}
! LDADD=	-lcrypt -lutil
  .PATH:  ${.CURDIR}/../../usr.bin/chpass ${.CURDIR}/../../usr.sbin/vipw \
! 	${.CURDIR}/../rlogin 
  
! CFLAGS+= -DLOGIN_CAP -DCRYPT -I. -I${.CURDIR} \
  	-I${.CURDIR}/../../usr.sbin/vipw \
  	-I${.CURDIR}/../../usr.bin/chpass \
  	-Dyp_error=warnx -DLOGGING
  
  CLEANFILES= ${GENSRCS}
