$FreeBSD$

*** inc-hist.texi.orig	Mon Mar 20 10:43:36 2000
--- inc-hist.texi	Mon Mar 20 10:48:17 2000
***************
*** 26,34 ****
  @node Using History Interactively
  @chapter Using History Interactively
  
! @ifclear BashFeatures
! @defcodeindex bt
! @end ifclear
  
  @ifset BashFeatures
  This chapter describes how to use the @sc{gnu} History Library
--- 26,34 ----
  @node Using History Interactively
  @chapter Using History Interactively
  
! @c @ifclear BashFeatures
! @c @defcodeindex bt
! @c @end ifclear
  
  @ifset BashFeatures
  This chapter describes how to use the @sc{gnu} History Library
***************
*** 39,47 ****
  @end ifset
  @ifclear BashFeatures
  This chapter describes how to use the GNU History Library interactively,
! from a user's standpoint.  It should be considered a user's guide.  For
! information on using the GNU History Library in your own programs,
! @pxref{Programming with GNU History}.
  @end ifclear
  
  @ifset BashFeatures
--- 39,48 ----
  @end ifset
  @ifclear BashFeatures
  This chapter describes how to use the GNU History Library interactively,
! from a user's standpoint.
! @c It should be considered a user's guide.  For
! @c information on using the GNU History Library in your own programs,
! @c @pxref{Programming with GNU History}.
  @end ifclear
  
  @ifset BashFeatures
