$FreeBSD$
Index: elf64-sparc.c
===================================================================
RCS file: /home/ncvs/src/contrib/binutils/bfd/elf64-sparc.c,v
retrieving revision 1.1.1.12
diff -u -u -0 -r1.1.1.12 elf64-sparc.c
--- elf64-sparc%FIXED.c	16 Jun 2004 05:44:32 -0000	1.1.1.12
+++ elf64-sparc%FIXED.c	17 Jun 2004 05:24:42 -0000
@@ -2790,2 +2790,2 @@
-      bfd_put_64 (output_bfd, (bfd_vma) 0,
-		  sgot->contents + (h->got.offset &~ (bfd_vma) 1));
+//XXX:DEO      bfd_put_64 (output_bfd, (bfd_vma) 0,
+//		  sgot->contents + (h->got.offset &~ (bfd_vma) 1));
